hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java

```diff
@@ -0,0 +1,171 @@
+
+    private final ConcurrentHashMap<Integer, SlowOperationLog> slowOperationLogs = new ConcurrentHashMap<Integer, SlowOperationLog>();
+
+    private long nowNanos;
+    private long nowMillis;
+
+    private volatile boolean running = true;
+
+    public SlowOperationDetector(Node node, BasicOperationScheduler scheduler) {
+        super(SlowOperationDetector.class.getSimpleName());
+        setDaemon(true);
+
+        genericOperationThreads = scheduler.genericOperationThreads;
+        partitionOperationThreads = scheduler.partitionOperationThreads;
+
+        genericCurrentOperationData = new CurrentOperationData[genericOperationThreads.length];
+        partitionCurrentOperationData = new CurrentOperationData[partitionOperationThreads.length];
+
+        initCurrentOperationData(genericCurrentOperationData);
+        initCurrentOperationData(partitionCurrentOperationData);
+
+        int thresholdMillis = Math.abs(node.groupProperties.SLOW_OPERATION_DETECTOR_THRESHOLD_MILLIS.getInteger());
+        int purgeIntervalSeconds = Math.abs(node.groupProperties.SLOW_OPERATION_DETECTOR_LOG_PURGE_INTERVAL_SECONDS.getInteger());
+        int logLifetimeSeconds = Math.abs(node.groupProperties.SLOW_OPERATION_DETECTOR_LOG_RETENTION_SECONDS.getInteger());
+
+        slowOperationThresholdNanos = TimeUnit.MILLISECONDS.toNanos(thresholdMillis);
+        logPurgeIntervalNanos = TimeUnit.SECONDS.toNanos(purgeIntervalSeconds);
+        logLifetimeNanos = TimeUnit.SECONDS.toNanos(logLifetimeSeconds);
+
+        logger = node.getLogger(BasicOperationScheduler.class);
+    }
+
+    @Override
+    public void run() {
+        long lastLogPurge = System.nanoTime();
+
+        while (running) {
+            nowNanos = System.nanoTime();
+            nowMillis = System.currentTimeMillis();
+
+            checkForSlowOperations(genericOperationThreads, genericCurrentOperationData);
+            checkForSlowOperations(partitionOperationThreads, partitionCurrentOperationData);
+
+            if (nowNanos - lastLogPurge > logPurgeIntervalNanos) {
+                lastLogPurge = nowNanos;
+                purgeLogs();
+            }
+
+            try {
+                TimeUnit.NANOSECONDS.sleep(ONE_SECOND_IN_NANOS - (System.nanoTime() - nowNanos));
+            } catch (InterruptedException ignored) {
+                break;
+            }
+        }
+    }
+
+    Collection<SlowOperationLog> getSlowOperationLogs() {
+        return slowOperationLogs.values();
+    }
+
+    public void shutdown() {
+        running = false;
+    }
+
+    private void initCurrentOperationData(CurrentOperationData[] currentOperationDataArray) {
+        for (int i = 0; i < currentOperationDataArray.length; i++) {
+            currentOperationDataArray[i] = new CurrentOperationData();
+            currentOperationDataArray[i].operationHashCode = -1;
+        }
+    }
+
+    private void checkForSlowOperations(BasicOperationScheduler.OperationThread[] operationThreads,
```

Comment:
Personally I would call this method 'scan' since the purpose of this class is to detect slow running operations anyway. I would also pull out the content of the loop into a seperate method. It is getting a bit on the long side.